core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -243,8 +249,14 @@ private LocalCheckpointTracker createLocalCheckpointTracker(
                 localCheckpoint = SequenceNumbers.NO_OPS_PERFORMED;
                 break;
             case OPEN_INDEX_AND_TRANSLOG:
+                // When recovering from a previous commit point, we use the local checkpoint from that commit,
+                // but the max_seqno from the last commit. This allows use to throw away stale operations.
```

Comment:
why do we need this distinction? can't we keep it conceptually simple? also we recover everything from the translog, so I'm not sure how it can happen that we throw away stuff?

Reply:
- If we use the local checkpoint from the last commit, operations having seqno less than or equal to the local checkpoint will be skipped -> we need to use the local checkpoint from the starting commit.

- We fillSeqNoGaps until the max_seqno -> we need to use max_seqno from the last commit to have full history.

Reply:
> We fillSeqNoGaps until the max_seqno -> we need to use max_seqno from the last commit to have full history.

We do so after we replay the translog. at which point we'll know what the max seq is (and it may be higher than the one in the last commit, but it can't be lower)